engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerService.java

```diff
@@ -0,0 +1,268 @@
+            public void close() throws IOException {
+                super.close();
+            }
+        };
+    }
+
+    private void stop(DockerClient dockerClient, String containerId) {
+        log.info("Stopping container: {}", containerId);
+        dockerClient.stopContainerCmd(containerId).exec();
+    }
+
+    private void remove(DockerClient dockerClient, String containerId) {
+        dockerClient.removeContainerCmd(containerId)
+                .withRemoveVolumes(true)
+                .exec();
+    }
+
+    public void pullImageIfNotExists(String imageName) {
+        try (DockerClient client = dockerClient()) {
+            List<String> available = listImages(client);
+            boolean exists = available.stream().anyMatch(n -> n.equals(imageName));
+            if (!exists) {
+                log.info("Image " + imageName + " not found locally");
+                pullImage(client, imageName);
+                log.info("Image " + imageName + " downloaded");
+            }
+        } catch (IOException ex) {
+            log.error(ex.toString());
+            throw new RuntimeException("Failed to pull image if not exists" + ex.getMessage(), ex);
+
+        }
+    }
+
+    public List<String> listImages(DockerClient client) {
+        List<Image> images = client.listImagesCmd().exec();
+        if (images == null) {
+            return Collections.emptyList();
+        }
+        return images.stream().map(Image::getRepoTags).filter(Objects::nonNull).flatMap(Arrays::stream).collect(Collectors.toList());
+    }
+
+    public void pullImage(DockerClient client, String imageName) {
+        log.info("Downloading " + imageName + " ... this may take some time, but we only need to do it once");
+        PullImageResultCallback callback = new PullImageResultCallback() {
+            @Override
+            public void onError(Throwable throwable) {
+                log.error("Failed to pull image" + throwable.getMessage());
+                super.onError(throwable);
+            }
+        };
+        try {
+            client.pullImageCmd(imageName).exec(callback).awaitCompletion();
+        } catch (InterruptedException ex) {
+            log.error(ex.toString());
+            throw new RuntimeException("Failed to pull image" + ex.getMessage(), ex);
+        }
+    }
+
+    public DockerClient dockerClient() {
+        DefaultDockerClientConfig.Builder builder = DefaultDockerClientConfig.createDefaultConfigBuilder()
+                .withDockerTlsVerify(false);
+        Optional.ofNullable(dockerProperties.getRegistryUrl()).ifPresent(builder::withRegistryUrl);
+        Optional.ofNullable(dockerProperties.getRegistryPassword()).ifPresent(builder::withRegistryPassword);
+        Optional.ofNullable(dockerProperties.getRegistryUsername()).ifPresent(builder::withRegistryUsername);
+        DefaultDockerClientConfig config = builder.build();
+
+        URI host = Optional.ofNullable(System.getProperty("use.socket.path")).map(path ->
```

Comment:
We configure application using environment variables. What is the reason to choose system property for this option?

Reply:
Mostly related to maintains consistency.